lambda/translquote.ml

```diff
@@ -0,0 +1,3144 @@
+      Pat.alias loc pat (Var.Value.mk (Lvar id))
+    | Tpat_constant const ->
+      let const = quote_constant loc const in
+      Pat.constant loc const
+    | Tpat_tuple pats ->
+      let pats =
+        List.map
+          (fun (lbl, p) -> quote_nonopt loc lbl, quote_value_pattern p)
+          pats
+      in
+      Pat.tuple loc pats
+    | Tpat_construct (lid, constr, args, _) ->
+      let constr = quote_constructor env lid.loc constr in
+      let args =
+        match args with
+        | [] -> None
+        | _ :: _ ->
+          let args = List.map quote_value_pattern args in
+          let with_labels =
+            List.map
+              (fun a -> Label.Nonoptional.no_label |> Label.Nonoptional.wrap, a)
+              args
+          in
+          Some (Pat.tuple loc with_labels |> Pat.wrap)
+      in
+      Pat.construct loc constr args
+    | Tpat_variant (variant, argo, _) ->
+      let argo = Option.map quote_value_pattern argo in
+      Pat.variant loc (Variant.of_string loc variant |> Variant.wrap) argo
+    | Tpat_record (lbl_pats, closed) ->
+      let lbl_pats =
+        List.map
+          (fun (lid, lbl_desc, pat) ->
+            let lbl = quote_record_field env Asttypes.(lid.loc) lbl_desc in
+            let pat = quote_value_pattern pat in
+            lbl, pat)
+          lbl_pats
+      in
+      let closed =
+        match closed with Asttypes.Closed -> true | Asttypes.Open -> false
+      in
+      Pat.record loc lbl_pats closed
+    | Tpat_array (_, _, pats) ->
+      let pats = List.map quote_value_pattern pats in
+      Pat.array loc pats
+    | Tpat_or (pat1, pat2, _) ->
+      let pat1 = quote_value_pattern pat1 in
+      let pat2 = quote_value_pattern pat2 in
+      Pat.or_ loc pat1 pat2
+    | Tpat_unboxed_tuple pats ->
+      let pats =
+        List.map
+          (fun (lbl, p, _) -> quote_nonopt loc lbl, quote_value_pattern p)
+          pats
+      in
+      Pat.unboxed_tuple loc pats
+    | Tpat_record_unboxed_product (lbl_pats, closed) ->
+      let lbl_pats =
+        List.map
+          (fun (lid, lbl_desc, pat) ->
+            let lbl = quote_record_field env Asttypes.(lid.loc) lbl_desc in
+            let pat = quote_value_pattern pat in
+            lbl, pat)
+          lbl_pats
+      in
+      let closed =
+        match closed with Asttypes.Closed -> true | Asttypes.Open -> false
+      in
+      Pat.unboxed_record loc lbl_pats closed
+    | Tpat_lazy pat ->
+      let pat = quote_value_pattern pat in
+      Pat.lazy_ loc pat
+  in
+  List.fold_right
+    (fun extra p -> quote_pat_extra loc p extra)
+    p.pat_extra (Pat.wrap pat_quoted)
+
+and quote_core_type ty =
+  let loc = ty.ctyp_loc in
+  match ty.ctyp_desc with
+  | Ttyp_var (None, _) -> Type.var loc None |> Type.wrap
+  | Ttyp_var (Some name, _) ->
+    let var =
+      match Hashtbl.find_opt vars_env.env_poly name with
+      | Some (_, var) -> var
+      | None ->
+        with_poly_type name;
```

Comment:
How confident are you these are correctly scoped? I don't even know where I'd expect to find the corresponding without.

Reply:
By "these", what are you referring to?